client/src/rubin/nublado/client/nubladoclient.py

```diff
@@ -82,27 +83,36 @@ async def __aiter__(self) -> AsyncIterator[SpawnProgressMessage]:
             Raised if a protocol error occurred while connecting to the
             EventStream API or reading or parsing a message from it.
         """
-        async for sse in self._source.aiter_sse():
-            try:
-                event_dict = sse.json()
-                event = SpawnProgressMessage(
-                    progress=event_dict["progress"],
-                    message=event_dict["message"],
-                    ready=event_dict.get("ready", False),
-                )
-            except Exception as e:
-                err = f"{type(e).__name__}: {e!s}"
-                msg = f"Error parsing progress event, ignoring: {err}"
-                self._logger.warning(msg, type=sse.event, data=sse.data)
-                continue
+        sse_events = self._source.aiter_sse()
+        try:
+            async for sse in sse_events:
+                try:
+                    event_dict = sse.json()
+                    event = SpawnProgressMessage(
+                        progress=event_dict["progress"],
+                        message=event_dict["message"],
+                        ready=event_dict.get("ready", False),
+                    )
+                except Exception as e:
+                    err = f"{type(e).__name__}: {e!s}"
+                    msg = f"Error parsing progress event, ignoring: {err}"
+                    self._logger.warning(msg, type=sse.event, data=sse.data)
+                    continue
 
-            # Log the event and yield it.
-            now = datetime.now(tz=UTC)
-            elapsed = int((now - self._start).total_seconds())
-            status = "complete" if event.ready else "in progress"
-            msg = f"Spawn {status} ({elapsed}s elapsed): {event.message}"
-            self._logger.info(msg, elapsed=elapsed, status=status)
-            yield event
+                # Log the event and yield it.
+                now = datetime.now(tz=UTC)
+                elapsed = int((now - self._start).total_seconds())
+                status = "complete" if event.ready else "in progress"
+                msg = f"Spawn {status} ({elapsed}s elapsed): {event.message}"
+                self._logger.info(msg, elapsed=elapsed, status=status)
+                yield event
+        finally:
+            # aiter_sse actually returns an asynchronous generator, which
+            # therefore is supposed to be explicitly closed with alcose()
+            # after break, unlike a true iterator. Handle this case to suppress
+            # warnings in Python 3.13.
+            with suppress(AttributeError):
```

Comment:
Why do we have to handle `AttributeError`s here (and elsewhere), instead of using `aclosing`?

Reply:
This is because the upstream typing says that they're returning an async iterator, but it's actually an async generator. Async iterators don't have `aclose` methods, and they could replace the generator with an iterator at any time and not break the interface guarantee. So we have to allow for the possibility that `aclose` doesn't exist, because it really is an iterator.

I've now refactored this into a helper function (which we could lift into Safir if we wanted to pursue this elsewhere) and added a bunch more documentation to make this clearer.

Reply:
Ohhhhhhh that's to suppress the case where `aclose` isn't there. I mistakenly thought it was to catch an `AttributeError` from the contents of `aclose`.  This makes total sense.

Reply:
Oh, yeah, the suppress might be a bit too broadly scoped. I didn't think about that. I suppose I should actually use `getattr` just to be safer?